Bonsai.Configuration/ScriptExtensions.cs

```diff
@@ -7,6 +7,7 @@
 using System.Reflection;
 using System.Xml;
 using System.Xml.Linq;
+using static Bonsai.Configuration.ScriptExtensionsProjectMetadata;
```

Comment:
@PathogenDavid refactoring looks great, and it made me wonder whether for consistency we should move the `ProjectFileTemplate` and `AddAssemblyReferences` method inside `ScriptExtensionsProjectMetadata`.

This would eliminate the need to make the constants internal and more cleanly separate out all metadata file manipulation.

To make this work cleanly we would probably need to change the signature to use package references instead of assembly references (to keep out the dependency on `PackageConfiguration`), move all file IO outside (including the call to `EnsureNuGetSettings`) and return a new instance of project metadata if we want to keep the struct strictly immutable.

Maybe something like this?

```c#
        public ScriptExtensionsProjectMetadata AddPackageReferences(IEnumerable<PackageReference> packageReferences)
        {
            var root = projectDocument?.Root;
            root ??= XElement.Parse(ProjectFileTemplate, LoadOptions.PreserveWhitespace);

            var projectReferences = root.Descendants(PackageReferenceElement).ToArray();
            var lastReference = projectReferences.LastOrDefault();
            foreach (var reference in packageReferences)
            {
                var includeAttribute = new XAttribute(PackageIncludeAttribute, reference.Id);
                var versionAttribute = new XAttribute(PackageVersionAttribute, reference.Version);
                var existingReference = (from element in projectReferences
                                         let id = element.Attribute(PackageIncludeAttribute)
                                         where id != null && id.Value == reference.Id
                                         select element).FirstOrDefault();
                if (existingReference != null)
                {
                    var version = existingReference.Attribute(PackageVersionAttribute);
                    if (version == null) existingReference.Add(versionAttribute);
                    else if (NuGetVersion.Parse(version.Value) < NuGetVersion.Parse(reference.Version))
                    {
                        version.SetValue(reference.Version);
                    }

                    continue;
                }

                var referenceElement = new XElement(PackageReferenceElement, includeAttribute, versionAttribute);
                if (lastReference == null)
                {
                    var itemGroup = new XElement(ItemGroupElement);
                    itemGroup.Add(Environment.NewLine + InnerIndent);
                    itemGroup.Add(referenceElement);
                    itemGroup.Add(Environment.NewLine + OuterIndent);

                    root.Add(OuterIndent);
                    root.Add(itemGroup);
                    root.Add(Environment.NewLine);
                    root.Add(Environment.NewLine);
                }
                else
                {
                    lastReference.AddAfterSelf(referenceElement);
                    if (lastReference.PreviousNode != null && lastReference.PreviousNode.NodeType >= XmlNodeType.Text)
                    {
                        lastReference.AddAfterSelf(lastReference.PreviousNode);
                    }
                }
                lastReference = referenceElement;
            }

            return new ScriptExtensionsProjectMetadata(new XDocument(root));
        }
```

Then the `AddAssemblyReferences` method itself on `ScriptExtensions` could be reduced to:

```c#
        public void AddAssemblyReferences(IEnumerable<string> assemblyReferences)
        {
            var projectMetadata = LoadProjectMetadata();
            if (!projectMetadata.Exists)
            {
                EnsureNuGetSettings();
            }

            var packageReferences = assemblyReferences
                .Select(assemblyName => packageConfiguration.GetAssemblyPackageReference(assemblyName, packageMap))
                .Where(package => package != null);

            projectMetadata = projectMetadata.AddPackageReferences(packageReferences);
            File.WriteAllText(ProjectFileName, projectMetadata.ToString());
        }
```

This assumes there is also a `ToString` override on `ScriptExtensionsProjectMetadata` to emit the formatted output. Below is a simple but probably error prone implementation:

```c#
        public override string ToString()
        {
            return projectDocument?.Root.ToString(SaveOptions.DisableFormatting);
        }
```

Reply:
> ```csharp
> .Where(package => package is not null);
> ```

I know the old implementation had this, but is this actually a good thing to do? This basically means we couldn't figure out how to reference the requested assembly, should we throw instead?

Reply:
Done!